Ux 469 <SidePanel />

nahumzs opened the pull request:
### 🛠 Purpose

close #79 

*The SidePanel is a Controlled Component*

## Props:

```js
const propTypes = {
  /** The content for the SidePanel. */
  children: PropTypes.node.isRequired,

  /** The width of the open panel. */
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /** Callback triggered when the side panel needs to be close */
  onClose: PropTypes.func,

  /** Callback once the sidepanel has been opened event */
  onAfterOpen: PropTypes.func,

  /** Callback once the sidepanel has been closed event */
  onAfterClose: PropTypes.func,

  /** Control the visibility of the side panel. This prop makes the side panel */
  isOpen: PropTypes.bool.isRequired,

  /** Control the z position of the sidepanel */
  zIndex: PropTypes.number,

  /** Control y offset of the sidepanel */
  offsetY: PropTypes.number,

  /** Modify the look of the SidePanel */
  kind: PropTypes.oneOf(["default", "child"]),

  /** Disable scroll overflow on document.body when SidePanel is open. */
  disableBodyOverflow: PropTypes.bool,
};
```

## Components
The sidepanel has the following component you can use to compose it:
```js
<SidePanel.Header />
<SidePanel.Footer />
<SidePanel.Overlay />
<SidePanel.FocusTrap />
<SidePanel.Trigger />
```

## The architecture

By default, the SidePanel render whatever you pass as children, but will filter 
`{ Header, Footer, Overlay, Trigger  and FocusTrap}` to properly render those components accordingly 
their optimal position, as an example the Trigger will be rendered as inline since is expected that 
99% of the cases will be a *Button*, while the `<List.Overlay />` and the `<SidePanel />` will be rendered under a `React.Portal`. 

Lastly `<ListBox.Header />` and `<ListBox.Footer />` will be at the top and bottom respectively. 

--- 

## The Ex.

```js
<SidePanel isOpen={isOpen} onEscKey={handleEvent}>
  <SidePanel.Trigger onClick={()=> handleEvent}>Open Sidepanel</SidePanel.Trigger>
   <SidePanel.Header onCloseClick={()=> handleEvent}>New SidePanel</SidePanel.Header>
   [...your_content...]
  <SidePanel.Footer isSticky onClickAccept={handleAccept}>
</SidePanel>
```

### With header and Overlay 

```js
<SidePanel isOpen={isOpen} onEscKey={handleEvent}>
  <SidePanel.Header onCloseClick={handleEvent} />
  <SidePanel.Overlay onClick={handleEvent} />
  [...your_content...]
</SidePanel>
```

---

### Sidepanel Groups
```js
import Sidepanel, { SidePanelGroup } from "@paprika/sidepanel"
<SidePanelGroup>
  <SidePanel></SidePanel>
  <SidePanel></SidePanel>
</SidePanelGroup>
```

## Accessibility

- Focus to the first focusable element accordingly to [wai](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal)
> - In all circumstances, focus moves to an element contained in the dialog.
> - Unless a condition where doing otherwise is advisable, focus is initially set on the first focusable element.
- The focus is trap under the side panel until is dismissed, I'm using [focus-trap-react](https://github.com/davidtheclark/focus-trap-react)
- wai `role=modal`
- If there is not Focusable element on the SidePanel focusable-react has been config to target the `<ListBox.Header />`

## Visual support
![2019-06-05 09 35 58](https://user-images.githubusercontent.com/19418590/58977145-f0ac5b80-877d-11e9-94dc-97f6cc14f321.gif)

![2019-06-05 12 37 32](https://user-images.githubusercontent.com/19418590/58985099-6d940100-878f-11e9-86e1-205a42f36bcb.gif)

### Scrolling NAV
![2019-06-05 12 38 06](https://user-images.githubusercontent.com/19418590/58985100-6d940100-878f-11e9-95f9-58eba1fe95ef.gif)


### Grouping Sidepanels

![2019-06-06 09 27 10](https://user-images.githubusercontent.com/19418590/59050043-3f212f00-883e-11e9-9ff7-c16be05f1c4d.gif)

### Footer
![2019-06-10 12 09 35](https://user-images.githubusercontent.com/19418590/59222408-474ed680-8b7e-11e9-8409-b5f80be50c85.gif)

### Focus Trap
![2019-06-10 12 10 11](https://user-images.githubusercontent.com/19418590/59222384-3c944180-8b7e-11e9-9f22-c55715ec4a0d.gif)

## Todo
- [x] *<ListBox.Trigger />*
- [x] *<ListBox.Header />*
- [x] ~<ListBox.Content />~ no need it.
- [x] *<ListBox.Footer />* `isSticky`
- [x] *<ListBox.FocusTrap />* let pass properties to `focus-trap-react`
- [x] *`<SidePanelStoryGroup />`* multiple sidepanel support
- [x] Header `kind={}` prop to allowed having different styles on the header
- [x] Testing SidePanel Jest
- [x] Testing SidePanel Cypress
- [ ] ~Imperative API~ `wont do` in this PR
- [x] SidePanel <SidePanel OffsetY={number} />  
- [x] Block the body from being scrollable


## NOTE 

`Fixed couple test annoying warnings after rebasing master.`

![Screen Shot 2019-06-10 at 1 03 20 PM](https://user-images.githubusercontent.com/19418590/59223614-36ec2b00-8b81-11e9-841e-cf0a9248fa9d.png)
![Screen Shot 2019-06-10 at 1 05 09 PM](https://user-images.githubusercontent.com/19418590/59223615-36ec2b00-8b81-11e9-9c1b-55345674a955.png)

AndreyChernykh commented:
@nahumzs 
1. What `<SidePanel.Trigger />` is for? And why it is a child of a side panel? 

2. Not exactly clear in what circumstances `onOpen` and `onClose` are called. 

3. Why does Side panel has `onEscKey` handler? I seems to me, that we can have a special component for that: 

```
state = {
 isOpen: true
}

return (
  <div> 
    <KeyListener key={Key.ESC} onPress={() => state.isOpen = false } />
    <SidePanel isOpen={state.isOpen} />
  </div>
)
```

nahumzs commented:
@AndreyChernykh 

- `<Trigger />` takes care of the focus automatically when you close the Sidepanel. So you don't have to worry about that. If you don't want to use the trigger (which is possible) the consumer will need to implement focus management manually. 

- `onOpen` and `onClose` are callback called once the animation of the side panel has finished. pretty much when `<div onAnimationEnd={} />` is triggered. 

- The reason for the onEscKey is `a11y`  requirements, modal required to implemented the `esc` button, while I like your idea of KeyListner, I'm not sure how we can "enforce" the consumer to implement the `esc` behaviour if it's not part of the Component. 

^ adding to your suggestion I think maybe for another version we can think about how to handle the `esc` key across several component `Modal, SidePanels, Takeover and Popover` and try to solve the issue where all "dialog" components are dismissed when the `esc key` event is fired, instead of only dismissed the latest one that was opened.

KirillKayumov commented on packages/SidePanel/SidePanel.stories.js:
```diff
@@ -0,0 +1,102 @@
+      hello lets open a sidepanel
+      <SidePanel isOpen={isOpen}>
+        <SidePanel.Trigger onClick={toggle}>{isOpen ? "close" : "open"}</SidePanel.Trigger>
+        <SidePanel.Overlay onClick={toggle} />
```
I have some concerns about the `SidePanel.Overlay` component:
1. An overlay is visually displayed not inside a `SidePanel`, but around it. I think it's not so obvious for a consumer to place `SidePanel.Overlay` as a child of `SidePanel` component.
2. If a consumer should render `SidePanel.Overlay` component manually, it automatically becomes "not required". Although looking at Projects module, there is only one `SidePanel` without an overlay (new To-do SidePanel). All the other side panels have it. Also, I believe that almost always a side panel should be closed if a user clicks on an overlay. Now a consumer should specify it twice: `<SidePanel onClose={this.handleClose}>` and `<SidePanel.Overlay onClick={this.handleClose}`. Maybe it's better to display an overlay by default and call `onClose` on overlay click. Then props for overlays can be passed like so:
```js
<SidePanel isOpen={isOpen} showOverlay={false}>...</SidePanel>
<SidePanel isOpen={isOpen} overlayBackground={...}>...</SidePanel>
OR
<SidePanel isOpen={isOpen} overlay={{ show: false }}>...</SidePanel>
<SidePanel isOpen={isOpen} overlay={{ background: "..." }}>...</SidePanel>
```
WDYT?

nahumzs commented:
@KirillKayumov 

### 1. I just implemented that if you click on the overlay fire the onClose, so no more specifying onClick on the `<SidePanel.Overlay />`. 💯 for the suggestion

### 2. About `Projects module` having more Sidepanel with overlay. 
While this is true, in incoming designs seems that designers are starting to use the sidepanels without overlay, also a lot of other modules like the ones handling data tent to use the one without the overlay. 

### 3. Agree, visually hierarchically the Overlay should be outside, the reason I extracted inside of the <Sidepanel /> is to remove a lot of properties from that level. I think that's is a good idea because let the `<SidePanel />` lean and with just a small amount of properties. And lets us add any number of props to the overlay without worrying about impacting the `<SidePanel />`

ex.
```js
<SidePanel isOpen={true} onClose={() => setState(isOpen: false)}>
    <SidePanel.Overlay />
    [...content]
</SidePanel>
```

alexzherdev commented:
$0.02:
- Having a separate `onEscKey` prop means accessibility is not baked in, we should avoid that. Even if we make the prop required, I've noticed folks are lazy to look into the console to check for errors.
- Related, I should only specify the close handler once somewhere (probably on the sidepanel itself), which should then be automatically propagated as the escape key handler, the overlay click handler, and the header's close button handler. Looks like you're almost there except the header.
- Speaking of closing, the current `onClose` on the sidepanel is more like an `onAfterClose`. When I see `onClose` I think of it as the callback where I'm supposed to make the (controlled) sidepanel close.
- `<Trigger>` doesn't seem like it's necessary? If it's just for focus handling then it looks like `focus-trap-react` [should be able to handle it](https://github.com/davidtheclark/focus-trap-react/blob/69c14950c427eddbee14550358ec516459cb0ba1/src/focus-trap-react.js#L64-L70)? And then we're not coupled to always having `<Button>` as the trigger.
- `<Content>` is not clear to me. The only thing it adds is padding. If you don't use it, you will get no padding. Are there use cases for both of these?
- Will we support sticky footers like we had in group permissions?
- Kind of conflicted between setting overlay as a prop vs as an element. I'm not sure if we'll ever need props on the overlay? And in our system in general as of now having overlays is more prevalent than not having them

nahumzs commented:
> Having a separate onEscKey prop means accessibility is not baked in, we should avoid that. Even if we make the prop required, I've noticed folks are lazy to look into the console to check for errors.

ohh I see, so what about if I change `onClose` to be the default handler when anything that can close the side panel is fire? Esc, click on overlay, header close button, etc, will call onClose. 

and then I will rename the current onClose to `onAfterClose`.

> <Trigger> doesn't seem like it's necessary? If it's just for focus handling then it looks like focus-trap-react should be able to handle it? And then we're not coupled to always having <Button> as the trigger.

Not too sure if react-trap returns the focus to the button, if it does it I will remove the Trigger. The Trigger is as well an optional Component, so you can use anything to open the sidePanel since the isOpen is controlled. 

> <Content> is not clear to me. The only thing it adds is padding. If you don't use it, you will get no padding. Are there use cases for both of these?

Just double check and seems none designer remember no having padding on the sidepanel so I guess we can remove this component.

> Will we support sticky footers like we had in group permissions?
Yes, will be supported, haven't started working on this.

> Kind of conflicted between setting overlay as a prop vs as an element. I'm not sure if we'll ever need props on the overlay? And in our system in general as of now having overlays is more prevalent than not having them

I'm ok with both alternatives if people feel like is better to have it on the props I'm ok 😃with it. Let's @mikrotron make the call on it.

tristanjasper commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,174 @@
+    "SidePanel.Overlay": OverlayExtracted,
+    "SidePanel.Header": HeaderExtracted,
+    "SidePanel.FocusTrap": FocusTrapExtracted,
+    children,
```
is this `children` meant to be here? 🤔

nahumzs replied:
yes, will include everything that has been pass as children except the extracted components.

nahumzs commented on packages/Heading/stories/examples/HeadingLevels.js:
```diff
@@ -80,12 +80,6 @@ const HeadingLevelsStory = () => {
       <Heading level={6} isLight>
         Heading Six with isLight
       </Heading>
-      <br />
```
This is adding propTypes validation errors. and no adding any benefit to the stories.

nahumzs commented on packages/SidePanel/SidePanel.cypress.js:
```diff
@@ -0,0 +1,81 @@
+  });
+
+  let count = 1;
+  it("should call onAfterOpen and onAfterClose", () => {
```
this can't because depends on `onAnimatedEnd` which is fired by css, so will not run in RTL.

nahumzs commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,193 @@
+};
+
+const defaultProps = {
+  isInline: false, // this is an internal prop use by SidePanelGroup (Group.js)
```
yes is not support it right now, I didn't want to stretch this PR to cover all cases like slideIn=[left and right], those can be add in incoming version. I tried to keep this PR smaller as possible for a first version.

nahumzs commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,193 @@
+
+  // Extracts
+  const {
+    "SidePanel.FocusTrap": FocusTrapExtracted,
```
yes, I added this one before the review you did in the ListBox I will fix it.

nahumzs commented on packages/SidePanel/src/components/Dialog/Dialog.styles.js:
```diff
@@ -0,0 +1,85 @@
+  from {
+    opacity: 0;
+    right: -${width};
+    transform: trasnlateX(100%);
```
😮

nahumzs commented on packages/Tokens/src/tokens.js:
```diff
@@ -48,10 +48,10 @@ const tokens = {
-    "purpleLighten20": "#b3a4d8",
-    "purpleLighten30": "#d1c7e8",
-    "purpleLighten40": "#efebf7",
+    "purpleLighten10": "#9884c5",
```
I think got versioned at some point and are installed when doing yarn because Lerna `pretrasnpile` create them. I will remove them from git history so the `.gitignore` can do their job.

nahumzs commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,193 @@
+import React from "react";
+import ReactDOM from "react-dom";
+import PropTypes from "prop-types";
+import { default as FocusTrapLibrary } from "focus-trap-react/dist/focus-trap-react"; // eslint-disable-line
```
https://github.com/acl-services/paprika/pull/84/files/decac2d7b3b1fea09cb5ecfdafff325938a422cd#diff-a70e4ad809affbf01bae992a555dc39eR10

I'm importing the shell component which has all the props with FocusTrap name, the shell component will let us build the prop table for documentation. You could argue that this one is the one that I should rename, but is minor IMO.

nahumzs commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,193 @@
+};
+
+const defaultProps = {
+  isInline: false, // this is an internal prop use by SidePanelGroup (Group.js)
```
why would want to expose the prop? Wouldn't that just increase the prop table with props that the consumer will not need? I'm doing it to improve readability.

nahumzs commented on cypress/support/commands.js:
```diff
@@ -9,3 +9,7 @@
   cy.visit(`http://localhost:9009/iframe.html?id=${story}`);
 });
+
+Cypress.Commands.add("getByTestId", str => {
```
seems like this is the correct approach. I don't see how to override getTestById and use our own `data-qa-anchor` apart from this

https://spectrum.chat/testing-library/general/data-test-id-vs-data-testid~4ba906b1-7390-4c6a-a22a-925011babe67?m=MTUyOTMzNTQ2OTAxNA==

https://github.com/testing-library/dom-testing-library/issues/136#issuecomment-432515243

I will add cypress-testing-library taking advantage of this conversation.

nahumzs commented on packages/SidePanel/SidePanel.cypress.js:
```diff
@@ -0,0 +1,81 @@
+    input.type(testString).should("have.value", testString);
+  });
+
+  it("should be possible to interact with multiple sidepanels in a group", () => {
```
I'm adding this as an integration test, just be 💯 that works in a browser scenario. that's all.

nahumzs commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,193 @@
+
+  // Hooks
+  const [isSidePanelMounted, setMount] = React.useState(props.isOpen);
+  const offsetScroll = useOffsetScroll(props.offsetY);
```
offsetY is the initial `top` position when the body can be scrollable the NAV will disappear so the SidePanel has to move up to fill the gap that the NAV has left.

you can see the example gif named: `Scrolling NAV` for a visual example.

nahumzs replied:
I can change the name of the variable if you think there is a more suitable name for it.

nahumzs commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,193 @@
+  const { onAfterClose, onAfterOpen, onClose, width, isInline, kind, offsetY, ...moreProps } = props;
+
+  // Hooks
+  const [isSidePanelMounted, setMount] = React.useState(props.isOpen);
```
more than for the initial mount is for when Closing it and the `onAnimationEnds`, I have to wait until the animation ends to be able to remove it, I have changed the name of the variable to `isSidePanelVisible` to be more clear in communicating the intention of the variable.

rauldudu replied:
should this be?

```
useEffect(() => setMount(props.isOpen), [props.isOpen]);
```

rauldudu replied:
Sorry I think the code I posted won't work but I wonder if you should consider `useEffect` anyway...

nahumzs replied:
@rauldudu  https://github.com/acl-services/paprika/pull/84/files#diff-a70e4ad809affbf01bae992a555dc39eR116 <-- I have one or do you mean something different?

nahumzs commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,193 @@
+    }
+  }
+
+  const trigger = TriggerExtracted ? React.cloneElement(TriggerExtracted, { ref: refTrigger }) : null;
```
I think for that specific case you can create your own `<Button />` and controlled everything with the `<SidePanel.FocusTrap />`. The intention of the Trigger is for a more "casual" easy way to consume this component.  If there is a need eventually I can figure out how to pass the ref eventually.

nahumzs commented on packages/SidePanel/src/components/Dialog/Dialog.js:
```diff
@@ -0,0 +1,74 @@
+      ref={refSidePanel}
+      {...moreProps}
+    >
+      {header ? React.cloneElement(header, { ...header.props, ref: refHeader, onClose }) : null}
```
🤔 why the would need a ref for the header? just curious to know if this is an edge case or is something that we really need to take in consideration?

nahumzs commented on packages/SidePanel/src/components/Dialog/Dialog.styles.js:
```diff
@@ -0,0 +1,85 @@
+
+const gapChildPanel = space * 7;
+const childPanel = `
+    height:calc(100% - ${gapChildPanel + 80}px);
```
I have added an eslint to fix all this issue.

nahumzs commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,193 @@
+    document.addEventListener("keydown", handleEscKey, false);
+
+    if (props.isOpen) {
+      setMount(true);
```
https://github.com/acl-services/paprika/pull/84#discussion_r296409909

nahumzs commented on packages/SidePanel/src/components/Group/Group.js:
```diff
@@ -0,0 +1,54 @@
+    document.body
+  );
+
+  return [sidePanelsGroup];
```
🧠💨

nahumzs commented on packages/SidePanel/src/components/FocusTrap/FocusTrap.js:
```diff
@@ -0,0 +1,37 @@
+
+// shell component for enhancing UXDeveloper
+export default function FocusTrap() {
+  return null;
```
I move the library into this one and now I'm just rendering <FocusTrap /> in SidePanel.js, leaving it out of Dialog, I can imagine that the Dialog could potentially be rendered as a regular sidebar which doesn't need a FocusTrap

nahumzs commented on packages/SidePanel/src/index.js:
```diff
@@ -0,0 +1,2 @@
+export { default as SidePanelGroup } from "./components/Group";
```
SidePanelGroup has multiple side panels and is not part of the SidePanel, but I guess the overlay is neither part of it. 

This actually was my first approach and then I didn't like it, but now that you are suggesting it maybe is the right approach, I will change it. 

<SidePanel.Group>
  <SidePanel></SidePanel>
  <SidePanel></SidePanel>
</SidePanel.Group>

nahumzs commented on packages/SidePanel/src/components/Header/Header.js:
```diff
@@ -0,0 +1,57 @@
+        {hasCloseButton ? (
+          <Button.Close
+            data-qa-anchor="sidepanel-header-close"
+            isSemantic={false}
```
because if I do it semantic some app in our platform will experience a "funny" bad style button because of the global styles.

nahumzs commented on packages/SidePanel/src/hooks/useOffsetScroll.js:
```diff
@@ -0,0 +1,25 @@
+    }
+  };
+
+  React.useLayoutEffect(() => {
```
During the advance hooks workshop, kent dodds demonstrated an [example](https://github.com/kentcdodds/advanced-react-hooks/blob/master/src/exercises/05.js#L8) pretty similar to this, he advice was if the effect will affect the visual element while interacting with the browser API, use Layout otherwise useEffect. 

That's the reason I'm using *useLayout*, also he even demonstrated that because react is really fast you can't perceive really a difference between useEffect vs useLayoutEffect in this case, that's the reason why he added a sleep function in [this line](https://github.com/kentcdodds/advanced-react-hooks/blob/master/src/exercises/05.js#L49) so the difference is more perceivable.

alexzherdev replied:
That example is different—it really does "affect the visual element" directly, while yours only sets a listener which has no (immediate) visual implications. In that case, `useEffect` is recommended since it allows React to schedule it better and reduce the workload in the commit phase.

nahumzs replied:
`useEffect`
![Screen Shot 2019-06-26 at 2 29 15 PM](https://user-images.githubusercontent.com/19418590/60216583-4cb74c80-981f-11e9-8ed5-281fa8888e88.png)

`useLayoutEffect`
![Screen Shot 2019-06-26 at 2 29 44 PM](https://user-images.githubusercontent.com/19418590/60216584-4cb74c80-981f-11e9-851b-efa04300d25e.png)

see how the Sidepanel is not a `0px` when scrolling? and in the second example when scrolling is actually at `0px`. 

Changing to useEffect will never update cypress on the scrolling event, useLayoutEffect occurs before useEffect in the cycle and is synchronous so will update the element and cypress can observe the changes.

alexzherdev replied:
I'd say that's probably a cypress issue then. But not going to block on this for sure

nahumzs commented on packages/SidePanel/src/hooks/useOffsetScrollFooter.js:
```diff
@@ -0,0 +1,35 @@
+  const [offsetY, setOffsetY] = React.useState(0);
+  const [opacity, setOpacity] = React.useState(1);
+
+  const handleOffset = debounce(() => {
```
The Sticky Footer definitely need more refinement to improve interaction experience, this debounce prevent to see the flickering happening all the time. I added a ticket so we can improve it later. https://github.com/acl-services/paprika/issues/97 I think for this first version the Footer is good enough.

nahumzs commented on packages/SidePanel/src/components/Header/Header.js:
```diff
@@ -0,0 +1,57 @@
+    hasCloseButton,
+    kind,
+    // injected by Dialog.js
+    onClose, // eslint-disable-line
```
I don't think the consumer will even care about this prop will be overridden by the component isn't?

alexzherdev replied:
Right, but for the developers working on this component we still want to make sure that the props are wired correctly

mikrotron commented on packages/SidePanel/package.json:
```diff
@@ -0,0 +1,32 @@
+  },
+  "dependencies": {
+    "@babel/runtime-corejs2": "^7.3.1",
+    "@paprika/heading": "^0.1.0",
```
`<SidePanel.Trigger>` and `<SidePanel.Header>` also depend on `<Button>`.

mikrotron commented on packages/SidePanel/readme.md:
````diff
@@ -0,0 +1,59 @@
+## Props:
+
+```js
+const propTypes = {
````
Duplicating the props here in the `README.md` seems like too much work... although a quick listing might be handy, with a link to the source code, [like this](https://github.com/acl-services/paprika/blob/master/packages/Button/README.md)? I'm not so sure yet what the best approach will be.

nahumzs replied:
npm provide a direct link from the npm landing page of the component, as long we use the url property under repository on the package.json file.

mikrotron commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,190 @@
+  children: PropTypes.node.isRequired,
+
+  /** The width of the open panel. */
+  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
```
Seems like alphabetic order will be the easiest standard to enforce, even if it's not always as logical.

mikrotron replied:
Also, I think we should definitely use semantic enums for `width`.  I think we should impose a limited number of options to enforce consistency with our design system.

nahumzs replied:
I thought about that, but I think the component shouldn't enforce the width from this side. Would make more sense to provide a specific token that the developer can use and pass them to this prop. `<SidePanel width={tokens.sidepanel.width.half} />` as an example. What do you think? 

We can discuss this maybe in on our slack channel?

nahumzs replied:
We can merge and then modify this after we agree in a pattern?

mikrotron replied:
Agreed.  This shouldn't block the whole component.  I think it's important to follow up and decide on a pattern soon though.

mikrotron commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,190 @@
+import { extractChildren } from "./helpers";
+import useOffsetScroll from "./hooks/useOffsetScroll";
+import useBodyOverflow from "./hooks/useBodyOverflow";
+import useEscapeKey from "./hooks/useEscapeKey";
```
Is one hook per file a rule?

nahumzs replied:
nop you can use multiple hooks in a file like:

const [user, setUser] = React.useState(null);
const [security, setSecurity] = React.useState(null);

mikrotron replied:
No, I mean the way `useEscapeKey` is the only hook exported from `hooks/useEscapeKey`.  This is preferred to `import { hook1, hook2 } from "./hooks/radicalHooks"`?

mikrotron commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,190 @@
+  kind: PropTypes.oneOf(["default", "child"]),
+
+  /** Disable scroll overflow on document.body when SidePanel is open. */
+  disableBodyOverflow: PropTypes.bool,
```
Is this necessary? Can we just _always_ hide the document scrollbars when an overlay is used?

nahumzs replied:
🤔 good point. Not 100% about this though, can you confirm if you want this or just a suggestion?

mikrotron replied:
I'd like to try it without.  It's easier to add a new prop later, than deprecate one.

mikrotron commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,190 @@
+  } = props;
+
+  // Hooks
+  const [isSidePanelVisible, setIsSidePanelVisible] = React.useState(props.isOpen);
```
Maybe just `isVisible`?

mikrotron commented on packages/SidePanel/src/components/Dialog/Dialog.js:
```diff
@@ -0,0 +1,78 @@
+
+  return (
+    <div
+      aria-modal="true"
```
Should this be dependent on `isInline` too? If the user can interact with the content below, it's not really a _modal_.

mikrotron commented on packages/SidePanel/src/SidePanel.js:
```diff
@@ -0,0 +1,190 @@
+      sidePanel = ReactDOM.createPortal(
+        <React.Fragment>
+          <FocusTrap focusTrapOptions={focusTrapOptions}>
+            <div>{dialog}</div>
```
Does it really need the `<div>`?

nahumzs replied:
yes it does, the FocusTrap needs to have only 1 element as children can't have two neither a `<Fragment />`

mikrotron commented on packages/SidePanel/src/components/Dialog/Dialog.js:
```diff
@@ -0,0 +1,78 @@
+  onAnimationEnd: PropTypes.func.isRequired,
+  onClose: PropTypes.func,
+  refHeader: PropTypes.shape({ current: PropTypes.instanceOf(Element) }).isRequired,
+  refSidePanelContent: PropTypes.shape({ current: PropTypes.instanceOf(Element) }).isRequired,
```
`refContent`? In general, including `SidePanel` in this component's names seems redundant, and shorter names will be easier to read.

nahumzs replied:
mmm actually is the ref for the SidePanel inside of the Dialog.js component. Makes sense to specify the name.

mikrotron commented on packages/SidePanel/src/hooks/useBodyOverflow.js:
```diff
@@ -0,0 +1,12 @@
+      return;
+    }
+
+    document.body.style.overflow = "auto";
```
I think I got into some trouble with the old SidePanel over this... We had a SidePanel in a Takeover... the Takeover already set the overflow to hidden, but when the SidePanel closed it would restore the overflow to auto and screw up the Takeover.  It's important to restore the overflow as it was, not assume it should be `auto`.

mikrotron replied:
https://github.com/acl-services/acl-ui/blob/master/src/components/SidePanel/SidePanel.js#L178

mikrotron commented on packages/SidePanel/src/components/Overlay/Overlay.js:
```diff
@@ -0,0 +1,45 @@
+  background: "#000",
+  hasOutsideClick: true,
+  onClose: null,
+  zIndex: 99,
```
`z(n)` where `n <= 7`.

nahumzs replied:
this doesn't work because z returns a full string `z-index: 1` I have added a new function under z named `z.toNumber(6)` which returns only the number that we need.

mikrotron commented on packages/SidePanel/src/components/Header/Header.styles.js:
```diff
@@ -0,0 +1,40 @@
+  color: ;
+  display: flex;
+  justify-content: space-between;
+  padding: 12px ${space}px; // creates a 48px height
```
How about:
```css
min-height: ${stylers.spacer(6)};
height: ${stylers.spacer(6)};
```

mikrotron commented on packages/SidePanel/src/components/Header/Header.styles.js:
```diff
@@ -0,0 +1,40 @@
+export const headerCSS = css`
+  border-bottom: 1px solid ${tokens.border.color};
+  box-sizing: border-box;
+  color: ;
```
✂️

nahumzs replied:
thank you! you found the damn null that I was seeing in the dev console and couldn't found! this makes my soul rest...

mikrotron commented on packages/SidePanel/src/components/Header/Header.styles.js:
```diff
@@ -0,0 +1,40 @@
+  .heading--level-4,
+  .heading--level-5 {
+    ${fontSize()};
+    font-weight: 400;
```
This concerns me. If we want to enforce the styling of the heading, maybe we can do it by just taking the content as a `string` and putting the `<Heading displayLevel={5} isLight>` component in ourselves.

I'm also starting to regret that we haven't implemented some kind of selector on the root node for each type of component... maybe not a `className`, but something would be convenient.

nahumzs replied:
That was my suggestion for `[data-component-type="listbox.header"]`

To me, that would super helpful you could target each component via jss, css or scss. as well in testing. even theme the components would be easier

mikrotron replied:
Yeah.  Let's make an issue to add these to all our components maybe? Then if there's a bigger conversation, it'll be separate from your PR.

nahumzs replied:
https://github.com/acl-services/paprika/issues/101 minor for this PR for now.

nahumzs replied:
btw about your previous comment? can we assume `displayLevel={5}` on the sidepanel? 🤔

nahumzs commented on packages/SidePanel/src/helpers.js:
```diff
@@ -0,0 +1,27 @@
+
+          components[child.type.componentType] = childs;
+        } else {
+          components[child.type.componentType] = child;
```
don't think so because I'm checking if the value of components['somevalue'] is an array. In case is not I want to keep just 1 component there instead of converting it into an array.

nahumzs commented on packages/SidePanel/src/hooks/useOffsetScrollFooter.js:
```diff
@@ -0,0 +1,35 @@
+  };
+
+  React.useEffect(() => {
+    if (ref.current) {
```
I tried this, but since ref.current is null at the beginning the effect will not get executed, passing ref.current allowed to listen for current and when change applies the effect.

mikrotron commented on packages/SidePanel/src/components/Header/Header.js:
```diff
@@ -0,0 +1,59 @@
+
+  return (
+    <div ref={ref} css={headerCSS} kind={kind} {...moreProps}>
+      <div tabIndex="-1">{props.children}</div>
```
I'm not going to suggest we change this now, but I might want to try extending this to be easier to consume.  `children` can remain in a raw container, but if props like `heading`, `onNext`, `onPrev` are provided, I think it would be nice to actually render with a `<Heading>` or `<PrevNextArrows>`.  What do you think?

nahumzs replied:
🤔 ohh you mean if I do something like

```
<SidePanel>
  <SidePanel.Header onNext onPrev>some stuff</SidePanel.Header>
</SidePanel> 
```

will render the `<PrevNextArrow />` that's a cool idea! yea totally out of scope for this PR

mikrotron commented on packages/SidePanel/src/components/Dialog/Dialog.styles.js:
```diff
@@ -0,0 +1,79 @@
+const childPanel = css`
+  border-bottom-left-radius: ${space / 2}px;
+  border-top-left-radius: ${space / 2}px;
+  height: calc(100% - ${gapChildPanel + 80}px);
```
What is `80`?

nahumzs replied:
To adjust the position of the child sidepanel so is not exactly at the same level of the Parent, I totally picked by `eye` there is not a hard guideline for this, later we can remove/add create a token if we need it.

mikrotron replied:
Oh, I see, theres a "gap" at the top AND bottom... and they're different... 🤨
Still, odd that one has a descriptive variable (`gapChildPanel `) and the other is a literal.

nahumzs replied:
🤷‍♂

mikrotron commented on packages/SidePanel/src/components/Dialog/Dialog.styles.js:
```diff
@@ -0,0 +1,79 @@
+
+const gapChildPanel = space * 7;
+const childPanel = css`
+  border-bottom-left-radius: ${space / 2}px;
```
`spaceSm`

mikrotron commented on packages/SidePanel/src/components/Dialog/Dialog.styles.js:
```diff
@@ -0,0 +1,79 @@
+`;
+
+export const dialogContentStyles = css`
+  padding: ${space * 2}px;
```
`stylers.spacer(2)`

mikrotron commented on packages/SidePanel/src/components/Dialog/Dialog.styles.js:
```diff
@@ -0,0 +1,79 @@
+import { css, keyframes } from "styled-components";
+import tokens from "@paprika/tokens/lib/tokens";
+
+const space = Number.parseInt(tokens.space, 10);
```
Shouldn't need this.

mikrotron commented on packages/SidePanel/src/components/Dialog/Dialog.styles.js:
```diff
@@ -0,0 +1,79 @@
+}
+
+export const dialogStyles = css`
+  background: #fff;
```
token

mikrotron commented on packages/Stylers/src/helpers.js:
```diff
@@ -27,10 +27,17 @@ export const lineHeight = (scale = 0) => `
+};
 
-  return validLevels.includes(level) ? `z-index: ${level};` : `z-index: 1;`;
+z.toNumber = level => {
```
Clever, but do you think it's better to be more consistent with other helpers and go with `z()` and `zValue()`?

https://github.com/acl-services/paprika/blob/master/packages/Stylers/src/helpers.js#L9-L13
https://github.com/acl-services/paprika/blob/master/packages/Stylers/src/helpers.js#L18-L22

nahumzs replied:
🤔 Value doesn't describe that you are getting a number you could be getting `1px` `1vh` `1%` `1` or `{1}` right?

mikrotron commented on packages/SidePanel/src/hooks/useFooterOffset.js:
```diff
@@ -0,0 +1,40 @@
+import React from "react";
+import debounce from "lodash.debounce";
+
+const useFooterOffset = (height, ref = null, isSticky) => {
```
This approach seems way more complicated than simply using a different (flexbox) layout when the `isSticky` prop is true.  No handlers, no debouncing... just flex-shrink, flex-grow, and overflow.

mikrotron replied:
Also... it still looks janky.  It floats and then reappears.  😕

nahumzs replied:
I added a comment about that the footer needs an upgrade, I added it as the last feature, this will not affect the API and is easier to refactor it in a new PR. 

I don't think we should stop the whole SidePanel only for the footer.

nahumzs replied:
I created an issue for this already https://github.com/acl-services/paprika/issues/97

mikrotron replied:
Okay, I agree.

alexzherdev replied:
Huh... I thought there was a specific reason we're not using flexbox here.

alexzherdev commented on package.json:
```diff
@@ -59,11 +59,12 @@
-    "@storybook/addon-links": "^5.1.9",
-    "@storybook/addons": "^5.1.9",
-    "@storybook/react": "^5.1.9",
+    "@storybook/addon-actions": "^5.0.11",
```
Hmm 😕
